src/lib/recurringPurposes.js

```diff
@@ -275,27 +386,99 @@ const getTimeserieWithPurpose = timeseries => {
   })
 }
 
-const isLoopTrip = timeserie => {
-  return (
-    timeserie?.aggregation?.startPlaceDisplayName ===
-    timeserie?.aggregation?.endPlaceDisplayName
-  )
-}
-
 export const findPurposeFromSimilarTimeserieAndWaybacks = async (
   client,
   timeserie
 ) => {
-  const similarTimeseries = await findSimilarTimeseries(client, timeserie)
+  const similarTimeseries = await findSimilarRecurringTimeseries(
+    client,
+    timeserie
+  )
+  log('info', `Found ${similarTimeseries.length} similar timeseries`)
+
   let tsWithPurpose = getTimeserieWithPurpose(similarTimeseries)
+
   if (!tsWithPurpose) {
     // if no similar trips with purpose found, try with waybacks trips
-    const waybackTimeseries = await findWaybackTimeseries(client, timeserie)
+    const waybackTimeseries = await findSimilarRecurringTimeseries(
+      client,
+      timeserie,
+      {
+        isWayBack: true
+      }
+    )
+    log('info', `Found ${waybackTimeseries.length} wayback timeseries`)
     tsWithPurpose = getTimeserieWithPurpose(waybackTimeseries)
   }
   return tsWithPurpose?.aggregation?.purpose || null
 }
 
+/**
+ * Try to set a purpose to a timeserie, from existing contacts' addresses
+ *
+ * It is based on the geo coordinates. If a match is found on the start or end
+ * address, a relationship is added, and the contact geo information is updated
+ * with the trip coordinates, to incrementally improve the geolocation precision.
+ *
+ * @param {object} client - The cozy client instance
+ * @param {object} timeserie - The timeserie to categorize
+ * @param {Array<object>} contacts - The contacts with geo information
+ * @returns {object} The timeserie to update
+ */
```

Comment:
I will not approve anymore a PR with object / array as params. It's not very useful. Can you update the type please?

Reply:
Well that's harsh :smile: 
I understand, but timeseries is typically the whole and big `io.cozy.timeries.geojson` doctype ; should I document it entirely? As well as contacts ?
I guess it would be actually a good practice, but this is not restricted to this service scope, the whole app should rely on it. 
Then, do we have a preferred method to make such doctype typing available for the whole app? Is this actually done elsewhere?

Reply:
> Well that's harsh 😄

I know, but we can make little step by little step. 

First, you can only create a typedef with an object in it.

Then, one day, we'll udpdate this typedef to be more precise. Again and again ;). 

> Then, do we have a preferred method to make such doctype typing available for the whole app?

You can create a types.js file, define the typedef there, and import it everywhere you need :)

Reply:
Hope you like it [c2d1507](https://github.com/cozy/coachCO2/pull/293/commits/c2d150766eafbbe3a5c3a9f5b8ca5368cc0fd804)